File: Searching Algorithms/binary_search.js
var n = 5,
  k = 5;
var arr = [1, 2, 3, 4, 5];

console.log(binarySearch(arr, n, k));

function binarySearch(arr, n, k) {
  var first = 0,
    last = n - 1;
  while (first <= last) {
    var mid = Math.floor((first + last) / 2);
    if (arr[mid] === k) return mid;
    else if (arr[mid] > k) last = mid - 1;
    else if (arr[mid] < k) first = mid + 1;
  }
  return -1;
}

File: Searching Algorithms/Binary_search.java
// Binary search is used to search a key element from multiple elements in an array.

class BinarySearch{
    public static void binarySearch(int arr[], int first, int last, int key){
        int mid = (first+last)/2;
        while (first <= last){
            
            if (arr[mid]<key){
                first = mid+1;
            }
            
            else if(arr[mid]==key){
                System.out.println("Element found at index: "+ mid);
                break;
            }
            
            else{
                last = mid-1;
            }
            
            mid = (first+last)/2;
        }
            
        if(first > last){
            System.out.println("Element not found");
        }
    }
        
        
    public static void main (String[] args) {
        int arr[] = {10,20,30,40,50};
        int key=40;
        int last = arr.length-1;
        binarySearch(arr, 0, last, key);
    }
    
}

// Time Complexity = O(logn)
// Space Complexity = O(1)
